Modix.Services/Moderation/MessageContentCheckBehaviour.cs

```diff
@@ -0,0 +1,134 @@
+                    message.Id, message.Author.Id);
+                return;
+            }
+
+            if (author.Id == _discordSocketClient.CurrentUser.Id)
+                return;
+
+            var isContentBlocked = await IsContentBlocked(channel, message);
+
+            if (!isContentBlocked)
+            {
+                return;
+            }
+
+            if (await _designatedChannelService.ChannelHasDesignationAsync(channel.Guild,
+                channel, DesignatedChannelType.Unmoderated, default))
+            {
+                return;
+            }
+
+            if (await _authorizationService.HasClaimsAsync(author, AuthorizationClaim.BypassMessageContentPatternCheck))
+            {
+                Log.Debug("Message {MessageId} was skipped because the author {Author} has the {Claim} claim",
+                    message.Id, message.Author.Id, AuthorizationClaim.BypassMessageContentPatternCheck);
+                return;
+            }
+
+            Log.Debug("Message {MessageId} is going to be deleted", message.Id);
+
+            await _moderationService.DeleteMessageAsync(message, "Unauthorized Message Content Link",
+                _discordSocketClient.CurrentUser.Id, default);
+
+            Log.Debug("Message {MessageId} was deleted because it contains blocked content", message.Id);
+
+            await message.Channel.SendMessageAsync($"Sorry {author.Mention} your link has been removed!");
+        }
+
+        private async Task<bool> IsContentBlocked(IGuildChannel channel, IMessage message)
+        {
+            var patterns = await _messageContentPatternService.GetPatterns(channel.GuildId);
+
+            foreach (var patternToCheck in patterns.Where(x => x.Type == MessageContentPatternType.Blocked))
+            {
+                // If the content is not blocked, we can just continue to check the next
+                // blocked pattern
+                if (!DoesContentMatchPattern(patternToCheck.Pattern))
+                {
+                    continue;
+                }
+
+                // if we find a block match, we need to check for any exemptions
+                foreach (var allowPattern in patterns.Where(x => x.Type == MessageContentPatternType.Allowed))
+                {
+                    if (DoesContentMatchPattern(allowPattern.Pattern))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+
+                bool DoesContentMatchPattern(string pattern)
+                {
+                    var regex = new Regex(
+                        pattern: pattern,
+                        options: RegexOptions.IgnoreCase,
+                        matchTimeout: TimeSpan.FromSeconds(2));
+
+                    var blockMatches = regex.Matches(message.Content);
+
+                    return blockMatches.Any();
+                }
+            }
```

Comment:
This logic will allow a message that contains blocked content, so long as it also contains allowed content in at least one other part of the message.

For example, if we block `discord.gg/robloxexploits` and allow `discord.gg/csharp`, this message will be allowed:
```
Visit discord.gg/robloxexploits and discord.gg/csharp!
```

Instead, I think we want to loop over the specific parts of the message that match the blocked regex and check whether they're allowed.

Reply:
We should add a unit test for this scenario as well

Reply:
Good catch - did not test this.